bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java

```diff
@@ -0,0 +1,333 @@
+     * high performance at runtime of the system, when the Rule Engine asks for any particular object, instead of
+     * waiting it for parsing every time.
+     * <p>
+     * The Map has for keys URLs of the files containing automation objects and for values - parsed objects.
+     */
+    protected Map<String, E> providedObjectsHolder = new HashMap<String, E>();
+
+    /**
+     * This map is used for mapping the imported automation objects to the file that contains them. This provides
+     * opportunity when an event for deletion of the file is received, how to recognize which objects are removed.
+     */
+    protected Map<URL, List<String>> providerPortfolio = new HashMap<URL, List<String>>();
+
+    /**
+     * This Map holds URL resources that waiting for a parser to be loaded.
+     */
+    protected Map<String, List<URL>> urls = new HashMap<String, List<URL>>();
+
+    protected List<ProviderChangeListener<E>> listeners = new ArrayList<ProviderChangeListener<E>>();
+
+    public void activate(Map<Parser<E>, Map<String, String>> parsers) {
+        for (Parser<E> parser : parsers.keySet()) {
+            addParser(parser, parsers.get(parser));
+        }
+        super.activate();
+        importResources(new File(WatchingDir));
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+        parsers.clear();
+        urls.clear();
+        providedObjectsHolder.clear();
+        listeners.clear();
+    }
+
+    @Override
+    protected String getSourcePath() {
+        return WatchingDir;
+    }
+
+    @Override
+    public void addProviderChangeListener(ProviderChangeListener<E> listener) {
+        synchronized (listeners) {
+            listeners.add(listener);
+        }
+
+    }
+
+    @Override
+    public Collection<E> getAll() {
+        synchronized (providedObjectsHolder) {
+            return providedObjectsHolder.values();
+        }
+    }
+
+    @Override
+    public void removeProviderChangeListener(ProviderChangeListener<E> listener) {
+        synchronized (listeners) {
+            listeners.remove(listener);
+        }
+    }
+
+    protected void addParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.put(parserType, parser);
+        List<URL> value = urls.get(parserType);
+        if (value != null && !value.isEmpty()) {
+            for (URL url : value) {
+                importFile(parserType, url);
+            }
+        }
+    }
+
+    protected void removeParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.remove(parserType);
+    }
+
+    protected void importResources(File file) {
+        if (file.exists()) {
+            File[] files = file.listFiles();
```

Comment:
I don't know, but need we to check if file is a directory (or perhaps a symlink to a directory) first?